### ISSUE TITLE
[REVIEW]: Native PDF Reader Library in Julia

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#1453]
**Submitting author:** @sambitdash (<a href="http://orcid.org/0000-0003-4856-7244">Sambit Kumar Dash</a>)
**Repository:** <a href="https://github.com/sambitdash/PDFIO.jl" target ="_blank">https://github.com/sambitdash/PDFIO.jl</a>
**Version:** v0.1.8
**Editor:** @alexhanna
**Reviewers:** @malmaud, @jarvist
**Archive:** <a href="https://doi.org/10.5281/zenodo.3541938" target="_blank">10.5281/zenodo.3541938</a>

## Status

[![status](http://joss.theoj.org/papers/742f48b0842cddf715b58a0bca2ffeb3/status.svg)](http://joss.theoj.org/papers/742f48b0842cddf715b58a0bca2ffeb3)

Status badge code:

```
HTML: <a href="http://joss.theoj.org/papers/742f48b0842cddf715b58a0bca2ffeb3"><img src="http://joss.theoj.org/papers/742f48b0842cddf715b58a0bca2ffeb3/status.svg"></a>
Markdown: [![status](http://joss.theoj.org/papers/742f48b0842cddf715b58a0bca2ffeb3/status.svg)](http://joss.theoj.org/papers/742f48b0842cddf715b58a0bca2ffeb3)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/sambitdash/PDFIO.jl" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@jarvist & @malmaud, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @alexhanna know.

✨ **Please try and complete your review in the next two weeks** ✨ 

### Conflict of interest
- [x] As the reviewer I confirm that I have read the [JOSS conflict of interest policy](https://github.com/openjournals/joss/blob/master/COI.md) and that there are no conflicts of interest for me to review this work.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/sambitdash/PDFIO.jl">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Version:** v0.1.8
- [x] **Authorship:** Has the submitting author (@sambitdash) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the function of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Authors:** Does the `paper.md` file include a list of authors with their affiliations?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **References:** Do all archival references that should have a DOI list one (e.g., papers, datasets, software)?

### Conflict of interest

- [x] As the reviewer I confirm that I have read the [JOSS conflict of interest policy](https://github.com/openjournals/joss/blob/master/COI.md) and that there are no conflicts of interest for me to review this work.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/sambitdash/PDFIO.jl">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Version:** v0.1.8
- [x] **Authorship:** Has the submitting author (@sambitdash) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the function of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Authors:** Does the `paper.md` file include a list of authors with their affiliations?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **References:** Do all archival references that should have a DOI list one (e.g., papers, datasets, software)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @jameshclrk, it looks like you're currently assigned as the reviewer for this paper :tada:.

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01453/joss.01453/10.21105.joss.01453.pdf)

**Comment by malmaud** [comment]
I've begun my review. 

In general, I think the need is clear - this package is a utility for reading PDF files in Julia, which is clearly useful to some people. 

The paper is intelligible, although suffers from grammatical errors throughout that should be fixed (as does the documentation). An example of a particularly ungrammatical sentence is  "Almost, all the functionality of PDF understanding is entirely written from scratch in Julia with only exception of usage of certain (de)compression codecs". 

I find the third paragraph ("The APIs are developed...") confusing. What is a "native script based language"? I also don't understand the meaning of the sentence "every well-developed native library out in the market need connectors". 

In "Their contribution can be viewed from the following link.", the URL of the link should be included in the paper - I don't think a JOSS paper should have embedded hyperlinks without the actual URL being visible in the document.

The paper should reference other Julia packages for handling PDF documents and explain how this package differs. In particular, I believe `Taro.jl` is the dominant solution at the moment.

At present, I don't think the documentation is sufficient. There are no examples or tutorials referenced in the readme or documentation. The only overview, the "API structure and design" section in `https://sambitdash.github.io/PDFIO.jl/stable/#API-Structure-and-Design-1`, is minimal. 

The documentation, as far as I've been able to find,  consists solely of a description of each of dozens of API method which has been autogenerated from the docstrings in the source code and is listed in alphabetical order. It is not reasonable to expect new users to read through that to get started, and I'm not personally willing to do so and so cannot personally test to see if this package is functional. 

The package did install, although I can't personally test its functionality since the documentation is too sparse for me to understand how to use it.

Running the automated test (`Pkg.test("PDFIO")`) fails for me (see https://github.com/sambitdash/PDFIO.jl/issues/62). From the logs, it looks like the issue is ultimately with a different downstream package (`ZipFile`), but nevertheless I can't verify this package passes its own tests.

In conclusion, I think major revisions to the documentation are needed, but if that can be accomplished, this software is a good fit for JOSS.

**Comment by sambitdash** [comment]
@malmaud Thanks for your review. I will update the documentation to address your concerns around the documentation. On the failing of the test cases, I guess for some reason the build step did not run when you added the package. If you can run the following command:

```
pkg> build PDFIO
```
the test should have passed successfully.

**Comment by malmaud** [comment]
OK, I just installed and tested in a fresh Julia environment and the tests pass now.

**Comment by labarba** [comment]
👋  @jameshclrk — We see no progress on your review checklist. Have you started on this? Do give us an updates.

@sambitdash — it looks like there remain several comments from @malmaud that need to be addressed. What's your status?

**Comment by sambitdash** [comment]
@labarba - I am going for a complete revamp of the documentation by the end of this month. Most of the suggested changes will be addressed as part of that. I will update when the task is accomplished.

**Comment by labarba** [comment]
@alexhanna Please follow up by email with @jameshclrk to know when we might expect their review.

**Comment by labarba** [comment]
@whedon remind @sambitdash in 3 weeks

**Comment by whedon** [comment]
Reminder set for @sambitdash in 3 weeks

**Comment by alexhanna** [comment]
@labarba Emailed @jameshclrk.

**Comment by sambitdash** [comment]
@whedon commands

**Comment by whedon** [comment]
Here are some things you can ask me to do:

```
# List Whedon's capabilities
@whedon commands

# List of editor GitHub usernames
@whedon list editors

# List of reviewers together with programming language preferences and domain expertise
@whedon list reviewers

EDITORIAL TASKS

# Compile the paper
@whedon generate pdf

# Compile the paper from alternative branch
@whedon generate pdf from branch custom-branch-name

# Ask Whedon to check the references for missing DOIs
@whedon check references

```

**Comment by sambitdash** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
PDF failed to compile for issue #1453 with the following error: 

 Error reading bibliography ./paper.bib (line 14, column 1):
unexpected "u"
expecting space, ",", white space or "}"
Error running filter pandoc-citeproc:
Filter returned error status 1
Looks like we failed to compile the PDF

**Comment by sambitdash** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01453/joss.01453/10.21105.joss.01453.pdf)

**Comment by sambitdash** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01453/joss.01453/10.21105.joss.01453.pdf)

**Comment by sambitdash** [comment]
Please review with the Version: v0.1.5 of the software.

**Comment by alexhanna** [comment]
@whedon set version as v0.1.5

**Comment by alexhanna** [comment]
@whedon set v0.1.5 as version

**Comment by whedon** [comment]
OK. v0.1.5 is the version.

**Comment by whedon** [comment]
:wave: @sambitdash, please update us on how things are progressing here.

**Comment by sambitdash** [comment]
I have updated as desired and waiting for comments now.

**Comment by labarba** [comment]
@alexhanna — did you ever hear back from @jameshclrk about this review? 
The author has made changes to the software, but the review remains inactive for more than a month now. Please have a look.

**Comment by labarba** [comment]
Yikes. It got 'round to my next rotation as AEiC and we've had no progress in this submission.
 @alexhanna — Could you take a new look here and give us a status update?

**Comment by alexhanna** [comment]
Ee, sorry about that @labarba. I haven't heard back from @jameshclrk. I emailed him on June 16. Should we remove him from the review and attempt to find another reviewer?

@malmaud -- can you have another pass at this package? Do the documentation changes address your concerns?

**Comment by malmaud** [comment]
I will take a look this week.

**Comment by alexhanna** [comment]
Hi @malmaud, it's been several weeks. Can you have another look at this?

**Comment by alexhanna** [comment]
:wave: @malmaud are you able to continue this review?
Removing @jameshclrk from this review, given non-response.
@vchuravy, @ahwillia, @ChrisRackauckas -- are any of you available to review this?

**Comment by alexhanna** [comment]
@whedon remove @jameshclrk as reviewer

**Comment by whedon** [comment]
OK, @jameshclrk is no longer a reviewer

**Comment by malmaud** [comment]
Yes, I’ll take another look today.

**Comment by malmaud** [comment]
OK, tested it out. Functionality seems to work - I was able to load a PDF of my own and extract the its text and query the metadata, and it all seemed accurate. I'm not a particular domain expert on PDF reader tools but it least seems to do what it promises to do in its README. 

I see examples have been added to most of the docstrings of the exported functions and an example is given on the README, which is great, as well as a welcome addition of a table of contents which organized the API. I still find that many of these docstrings have some spelling or grammatical errors, but it's good enough to figure out how to use the package.

I think the corresponding JOSS paper still needs a bit of work, though - there are quite a few remaining grammar and punctuation issues which make it difficult to read. I also find it pretty hard to understand the claims being made. I don't understand the comments about MVC architecture in benefit number 2, and benefit 3 talks about the ability to modify PDFs, although the package doesn't actually support that at the moment. I don't understand in downside 1 what it means for a handle to be 'merely representational'. 

I think if the paper can be made more clear and some of the more nebulous claims and phrases rectified, this should be in good shape though.

**Comment by sambitdash** [comment]
Thanks @malmaud for your review. 

I have modified the paper based on your inputs.

**Comment by sambitdash** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
PDF failed to compile for issue #1453 with the following error: 

 /app/vendor/ruby-2.4.4/lib/ruby/2.4.0/psych.rb:377:in `parse': (tmp/1453/paper/paper.md): mapping values are not allowed in this context at line 2 column 13 (Psych::SyntaxError)
	from /app/vendor/ruby-2.4.4/lib/ruby/2.4.0/psych.rb:377:in `parse_stream'
	from /app/vendor/ruby-2.4.4/lib/ruby/2.4.0/psych.rb:325:in `parse'
	from /app/vendor/ruby-2.4.4/lib/ruby/2.4.0/psych.rb:252:in `load'
	from /app/vendor/ruby-2.4.4/lib/ruby/2.4.0/psych.rb:473:in `block in load_file'
	from /app/vendor/ruby-2.4.4/lib/ruby/2.4.0/psych.rb:472:in `open'
	from /app/vendor/ruby-2.4.4/lib/ruby/2.4.0/psych.rb:472:in `load_file'
	from /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-1e4ee47b240d/lib/whedon.rb:115:in `load_yaml'
	from /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-1e4ee47b240d/lib/whedon.rb:85:in `initialize'
	from /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-1e4ee47b240d/lib/whedon/processor.rb:36:in `new'
	from /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-1e4ee47b240d/lib/whedon/processor.rb:36:in `set_paper'
	from /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-1e4ee47b240d/bin/whedon:55:in `prepare'
	from /app/vendor/bundle/ruby/2.4.0/gems/thor-0.20.3/lib/thor/command.rb:27:in `run'
	from /app/vendor/bundle/ruby/2.4.0/gems/thor-0.20.3/lib/thor/invocation.rb:126:in `invoke_command'
	from /app/vendor/bundle/ruby/2.4.0/gems/thor-0.20.3/lib/thor.rb:387:in `dispatch'
	from /app/vendor/bundle/ruby/2.4.0/gems/thor-0.20.3/lib/thor/base.rb:466:in `start'
	from /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-1e4ee47b240d/bin/whedon:116:in `<top (required)>'
	from /app/vendor/bundle/ruby/2.4.0/bin/whedon:23:in `load'
	from /app/vendor/bundle/ruby/2.4.0/bin/whedon:23:in `<main>'

**Comment by sambitdash** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01453/joss.01453/10.21105.joss.01453.pdf)

**Comment by malmaud** [comment]
Great, much improved! I'd be happy with accepting in this state.

**Comment by alexhanna** [comment]
Thanks so much @malmaud. Can you confirm this by checking off the necessary boxes in your review?

**Comment by malmaud** [comment]
done

**Comment by sambitdash** [comment]
Thanks @malmaud.

**Comment by alexhanna** [comment]
:wave: @jarvist, @Datseris, @dmbates -- we're looking for a second reviewer for this piece after our first one bowed out. Could you possibly have a look at this?

**Comment by Datseris** [comment]
Thank you for the invitation, but I am sorry, I have to decline, too many deadlines to meet.

**Comment by jarvist** [comment]
I am on holiday until the 28th of October, but can then review this.

**Comment by alexhanna** [comment]
@whedon remove @jameshclrk as reviewer

**Comment by jarvist** [comment]
> Hi @jarvist -- are you able to review this now?

Certainly!

**Comment by alexhanna** [comment]
@whedon add @jarvist as reviewer

**Comment by whedon** [comment]
OK, @jarvist is now a reviewer

**Comment by jarvist** [comment]
I've just been reviewing this software + paper, but I can't seem to edit the checkboxes (also, they are still allocated to James Clark. @alexhanna - is this something you can fix?

Generally I'm happy with the package... it installs and tests run. I think the latest draft of the paper looks good. 

However, I still think the package README.md could be edited down to make it more clear, and make it easier for a user to understand what the functionality is, and to get started. 
For instance, you can:-
`julia> pdPageExtractText(stdout, pdDocGetPage(pdDocOpen("10.21105.joss.01453.pdf"), 1))` 
from the REPL to extract the first page of the JOSS review into the terminal. This is pretty cool functionality, but I had to find this out by fiddling around with it.

Are there any other high level access functions than `pdPageExtractText` ? 

From the point of view of the Package/API (not having read the source code), the namespace is quite confusing: the package is `PDFIO` but then all the exported functions are prepended with `pd` which seems superfluous for Julia, because everything is already under the `PDFIO` namespace of the module.
Similarly, the function arguments seem to vary quite a bit in placing and expectation. Does pdPageExtractText need an io object as the first argument? A default if single argument to return the text object would make a lot of sense. (Making the `stdout` in my example above superfluous.)
Also, the library seems to fail on Google-docs exported PDFs. (My first test case.) I will file an issue.

**Comment by alexhanna** [comment]
Hrm, try now @jarvist? I changed the name on the issue, but I'm not sure if that does anything.

**Comment by jarvist** [comment]
Still no 'edit' option in the '...' drop down menu.

**Comment by alexhanna** [comment]
You shouldn't need to edit it. You can click the checkboxes without editing. Does that work for you?

**Comment by jarvist** [comment]
No joy there either. (I think the GUI checkbox clicking is predicated by edit permissions.)

**Comment by sambitdash** [comment]
Hi @jarvist,

Thanks a lot for taking time to review the software and the paper. Also reporting the bug on the product. 
> 
> Are there any other high level access functions than `pdPageExtractText` ?

All methods starting with `pd` documented under: https://sambitdash.github.io/PDFIO.jl/dev/#PD-1 are high level functions. 

> 
> From the point of view of the Package/API (not having read the source code), the namespace is quite confusing: the package is `PDFIO` but then all the exported functions are prepended with `pd` which seems superfluous for Julia, because everything is already under the `PDFIO` namespace of the module.

`pd` does not stand for `PDFIO` but `PDF Document` or simply called `PD Layer` in most PDF format architectures. This is to differentiate from low level functions called `COS Layer` which requires you to understand the PDF objects like dictionary, arrays to be able to understand the PDF file. An explanation of these can be seen at: https://sambitdash.github.io/PDFIO.jl/dev/arch/

Moreover, the similar nomenclature is used by other established PDF libraries in the market. Hence, people who have used other PDF libraries may quickly understand the scope of the APIs. 

> Similarly, the function arguments seem to vary quite a bit in placing and expectation. Does pdPageExtractText need an io object as the first argument? A default if single argument to return the text object would make a lot of sense. (Making the `stdout` in my example above superfluous.)

1. PDF files can be large expanding up to several megabytes and can technically grow to gigabytes. With extract text returning strings this may lead to several copies of large strings. IO provides a cleaner interface for the same. Many a times users may extract all pages and `IO` provides a possibility to append pages over one another. 
2. PDF being an encrypted file format immutable extracted text should not be left in the memory which is a definite possibility with usage of strings. IO can be used to immediately write to a file or clear up the IOBuffer at the end of using the data. 

> Also, the library seems to fail on Google-docs exported PDFs. (My first test case.) I will file an issue.

This has been fixed. If you think this is a gating requirement for the paper, please let me know so that I can release a new version of the software.

**Comment by alexhanna** [comment]
👋 @openjournals/joss-eics can you look into the checkbox issue for new reviewers?

**Comment by danielskatz** [comment]
@arfon - can you help with the reviewer invitation here?

**Comment by arfon** [comment]
@jarvist - you've been invited by GitHub as a collaborator on this repository - you just need to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

**Comment by jarvist** [comment]
Strangely I was listed as a collaborator already (from a previous review?); but could not edit. All sorted now, thank you!

**Comment by jarvist** [comment]
> This has been fixed. If you think this is a gating requirement for the paper, please let me know so that I can release a new version of the software.

I don't think this is a gating requirement.

However, I do think the README.md should be edited to make it more clear. 

- Some kind of interactive example would really help: I had to read and process your function given in the illustration, to slowly understand enough to be able to play with it on the REPL. 

- Merging the `illustration` and `Sample Code` sections might make sense. Most on the non-code in `Sample Code` I think would be clearer as in-line comments within the code.

- I don't think the references to `Taro.jl` are necessary in the main text, instead they should be part of a `Alternative Software` or similar section much further down.

**Comment by sambitdash** [comment]
@jarvist updated as suggested.

**Comment by jarvist** [comment]
That's great! I think it's much improved, and am happy to sign off the full review.

**Comment by jarvist** [comment]
@alexhanna , I think we're good to go here, when you have time.

**Comment by sambitdash** [comment]
@alexhanna, Kindly let me know if there is anything needed from my side to bring this issue to a closure.

**Comment by alexhanna** [comment]
Hi @sambitdash - we need you to do is get a DOI number from Zenodo and post it here. We can then assign the repository a DOI, give a final proofread, and pass it off to the EIC for acceptance.

**Comment by alexhanna** [comment]
@whedon check references

**Comment by whedon** [comment]
```
Attempting to check references...
```

**Comment by whedon** [comment]
```Reference check summary:

OK DOIs

- None

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by alexhanna** [comment]
@whedeon check pdf

**Comment by alexhanna** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01453/joss.01453/10.21105.joss.01453.pdf)

**Comment by alexhanna** [comment]
I've also added a few changes to the paper text in https://github.com/sambitdash/PDFIO.jl/pull/77.

**Comment by alexhanna** [comment]
Thanks for merging those, @sambitdash.
@openjournals/joss-eics this one is good to go.

**Comment by sambitdash** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01453/joss.01453/10.21105.joss.01453.pdf)

**Comment by sambitdash** [comment]
@alexhanna 10.5281/zenodo.3541938 is the DoI reserved. Do I publish the article on Zenodo?

**Comment by sambitdash** [comment]
Uploaded the software and Zenodo with the stated doi.

**Comment by labarba** [comment]
The latest release in the repository is v0.1.8 — is that the reviewed version? Should we update the version number here?

**Comment by sambitdash** [comment]
The version number can be updated to 0.1.8. The latest update to the README is incorporated in it.

**Comment by labarba** [comment]
@whedon set v0.1.8 as version

**Comment by whedon** [comment]
OK. v0.1.8 is the version.

**Comment by labarba** [comment]
@whedon set 10.5281/zenodo.3541938 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.3541938" target="_blank">10.5281/zenodo.3541938</a> is the archive.

**Comment by labarba** [comment]
@whedon accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
```Reference check summary:

OK DOIs

- None

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/1099

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/1099, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by labarba** [comment]
@whedon accept deposit=true

**Comment by whedon** [comment]
```
Doing it live! Attempting automated processing of paper acceptance...
```

**Comment by whedon** [comment]
🐦🐦🐦 👉 [Tweet for this paper](https://twitter.com/JOSS_TheOJ/status/1195036068711714818) 👈 🐦🐦🐦

**Comment by whedon** [comment]
🚨🚨🚨 **THIS IS NOT A DRILL, YOU HAVE JUST ACCEPTED A PAPER INTO JOSS!** 🚨🚨🚨

 Here's what you must now do:

0. Check final PDF and Crossref metadata that was deposited :point_right: https://github.com/openjournals/joss-papers/pull/1100
1. Wait a couple of minutes to verify that the paper DOI resolves [https://doi.org/10.21105/joss.01453](https://doi.org/10.21105/joss.01453)
2. If everything looks good, then close this review issue.
3. Party like you just published a paper! 🎉🌈🦄💃👻🤘

 Any issues? notify your editorial technical team...

**Comment by labarba** [comment]
Congratulations, @sambitdash, your JOSS paper is published! 🚀 

Many thanks to our editor: @alexhanna, and the reviewers: @malmaud, @jarvist — your contributions to JOSS are greatly appreciated 🙏

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippets:

```
Markdown:
[![DOI](https://joss.theoj.org/papers/10.21105/joss.01453/status.svg)](https://doi.org/10.21105/joss.01453)

HTML:
<a style="border-width:0" href="https://doi.org/10.21105/joss.01453">
  <img src="https://joss.theoj.org/papers/10.21105/joss.01453/status.svg" alt="DOI badge" >
</a>

reStructuredText:
.. image:: https://joss.theoj.org/papers/10.21105/joss.01453/status.svg
   :target: https://doi.org/10.21105/joss.01453
```
This is how it will look in your documentation:

[![DOI](https://joss.theoj.org/papers/10.21105/joss.01453/status.svg)](https://doi.org/10.21105/joss.01453)

**We need your help!**

Journal of Open Source Software is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future. You can add your name to the reviewer list here: https://joss.theoj.org/reviewer-signup.html
- Making a small donation to support our running costs here: https://numfocus.org/donate-to-joss

**Comment by sambitdash** [comment]
Thanks @alexhanna, @labarba, @malmaud and @jarvist for bringing this submission to a successful closure.
